### PR TITLE
Remove ArrayVec in favor of a simple array

### LINKED ISSUE #31: Replace ArrayVec in ConnectionId with a simple array
This is what I had originally in quinn. This would get rid of the extra dependency, and it would let `ConnectionId` implement `Copy`, which would obviate the need for cloning.

### PR DESCRIPTION
**Pull request by twilco**
Closes #31.

### CODE REVIEW

**Comment by twilco** [comment]
Hi @djc!  I was able to remove `ArrayVec`, but haven't been able to get the tests to pass after a bit of digging.  Could you help point me in the right direction?  Let's take the `stop_stream` failure as an example:

```
---- tests::stop_stream stdout ----
INFO connecting
DEBG dropping packet from unrecognized connection, side: Client, header: Long { ty: 125, source_id: ConnectionId { len: 18, bytes: [236, 31, 120, 215, 219, 214, 88, 142, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] }, destination_id: ConnectionId { len: 18, bytes: [160, 191, 201, 164, 234, 127, 204, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] }, number: 738246780 }
```
"dropping packet from unrecognized connection" is clearly a bad sign.  Something seems to be up with how `Endpoint`s `connection_ids`, `connection_ids_initial`, and `connection_remotes` are being populated, but I haven't been able to determine what.  Is there anything that jumps out to you in this diff that could cause this?

**Comment by twilco** [comment]
Thanks @Ralith!  I see why my previous implementation was wrong.  That indeed fixed one test - now eight to go.  The ones failing all look similar to that of the one above.

**Comment by Ralith** [comment]
The `Eq`, `PartialEq`, `Ord`, `PartialOrd`, and `Hash` implementations should probably have the same fix applied.

File: quinn-proto/src/packet.rs [hunk 1]
```diff
@@ -271,8 +270,8 @@ impl Packet {
                         "destination connection ID longer than packet",
                     ));
                 }
-                buf.copy_to_slice(&mut cid_stage[0..dest_id_len]);
-                let id = ConnectionId::new(cid_stage, dest_id_len);
+                buf.copy_to_slice(&mut cid_stage[..dest_id_len]);
+                let id = ConnectionId::new(&cid_stage);
```
Comment (Ralith) [inline quinn-proto/src/packet.rs]: s/`&cid_stage`/`&cid_stage[..dest_id_len]`/

File: quinn-proto/src/packet.rs [hunk 2]
```diff
@@ -227,9 +226,9 @@ impl Packet {
                     ));
                 }
                 buf.copy_to_slice(&mut cid_stage[0..dcil as usize]);
-                let destination_id = ConnectionId::new(cid_stage, dcil as usize);
+                let destination_id = ConnectionId::new(&cid_stage);
                 buf.copy_to_slice(&mut cid_stage[0..scil as usize]);
-                let source_id = ConnectionId::new(cid_stage, scil as usize);
+                let source_id = ConnectionId::new(&cid_stage);
```
Comment (Ralith) [inline quinn-proto/src/packet.rs]: s/`&cid_stage`/`&cid_stage[..scil as usize]`/

File: quinn-proto/src/packet.rs [hunk 3]
```diff
@@ -227,9 +226,9 @@ impl Packet {
                     ));
                 }
                 buf.copy_to_slice(&mut cid_stage[0..dcil as usize]);
-                let destination_id = ConnectionId::new(cid_stage, dcil as usize);
+                let destination_id = ConnectionId::new(&cid_stage);
```
Comment (Ralith) [inline quinn-proto/src/packet.rs]: s/`&cid_stage`/`&cid_stage[..dcil as usize]`/

File: quinn-proto/src/packet.rs [hunk 4]
```diff
@@ -309,41 +308,53 @@ impl Packet {
...
+        rng.fill_bytes(&mut rng_bytes);
+        (&mut res.bytes[..len as usize]).clone_from_slice(&rng_bytes[..len as usize]);
+        res
     }
 }
 
 impl fmt::Display for ConnectionId {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for byte in &self.0 {
+        for byte in &self.bytes {
```
Comment (Ralith) [inline quinn-proto/src/packet.rs]: This, too, should be bounded by `len`, perhaps by dispatching to the `Deref` impl if that's convenient.

**Comment by twilco** [comment]
And we've got a ✅.  Mind taking another look, @Ralith?

File: quinn-proto/src/crypto.rs [hunk 5]
```diff
@@ -473,12 +473,8 @@ mod test {
...
     #[test]
     fn key_derivation() {
-        let id = ConnectionId(
-            [0x83, 0x94, 0xc8, 0xf0, 0x3e, 0x51, 0x57, 0x08]
-                .iter()
-                .cloned()
-                .collect(),
-        );
+        let data = [0x83, 0x94, 0xc8, 0xf0, 0x3e, 0x51, 0x57, 0x08];
+        let id = ConnectionId::new(&data);
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: If `data` isn't needed elsewhere, this should probably just be a oneliner.

File: quinn-proto/src/endpoint.rs [hunk 6]
```diff
@@ -317,7 +317,6 @@ impl Endpoint {
         //
         // Handle packet on existing connection, if any
         //
-
```
Comment (Ralith) [inline quinn-proto/src/endpoint.rs]: Accidental deletion?

File: quinn-proto/src/packet.rs [hunk 7]
```diff
@@ -309,41 +308,53 @@ impl Packet {
...
-        let mut x = ConnectionId(data.into());
-        x.0.truncate(len);
-        x
+    pub fn new(bytes: &[u8]) -> Self {
+        debug_assert!(bytes.is_empty() || (bytes.len() > 3 && bytes.len() <= MAX_CID_SIZE));
+        let mut res = Self {
+            len: bytes.len() as u8,
+            bytes: [0; MAX_CID_SIZE],
+        };
+        (&mut res.bytes[..bytes.len()]).clone_from_slice(&bytes);
```
Comment (Ralith) [inline quinn-proto/src/packet.rs]: Is the `(&mut ...)` trick really necessary?

File: quinn-proto/src/packet.rs [hunk 8]
```diff
@@ -309,41 +308,53 @@ impl Packet {
...
+        rng.fill_bytes(&mut rng_bytes);
+        (&mut res.bytes[..len as usize]).clone_from_slice(&rng_bytes[..len as usize]);
+        res
     }
 }
 
 impl fmt::Display for ConnectionId {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for byte in &self.0 {
+        for byte in (*(&self)).iter() {
```
Comment (Ralith) [inline quinn-proto/src/packet.rs]: Could this be written `self.iter()`?

**Comment by djc** [comment]
Also let's squash all these commits before merging?

**Comment by djc** [comment]
This looks great!

**Comment by twilco** [comment]
Thank you both for your help - I learned a lot!

**Comment by djc** [comment]
@twilco would you like to dig in a little more with another issue?

**Comment by twilco** [comment]
@djc, sure, I'd love to.  What did you have in mind?

**Comment by djc** [comment]
I've just written up some notes in #51, have a look to see if that's something you're interested in.

**Pull request merged**